blob/gcsblob/gcsblob.go

```diff
@@ -227,15 +251,80 @@ func OpenBucket(ctx context.Context, client *gcp.HTTPClient, bucketName string,
 	if err != nil {
 		return nil, err
 	}
+	otherSignerAvailable := len(drv.opts.PrivateKey) >= 1 || drv.opts.SignBytes != nil
```

Comment:
Assuming the above comments about whether this is necessary are resolved as "yes":

Could this be implemented instead by setting `Options.SignBytes` during initialization? Then again, you don't need to modify any of the code here or below.

Reply:
This would move the required fields into `gcsblob.Bucket` (going with the default style to prefer struct members over closures), and/or make the lazy initialization no longer a lazy one (the *Credentials API client* connects right away).

Reply:
Could you make a new struct with whatever fields you need, including a lazily initialized client, and pass a method of that struct as the `Options.SignBytes` func? I think it's OK to not close the client on shutdown, so `Bucket` and `bucket` wouldn't need to know about the struct at all.

Reply:
The motivation for this suggestion is to keep the number of ways we sign the same instead of adding a new one; it keeps the main code simpler. It also provides a demonstration of how to use `SignBytes`; if a user wants to modify how the default `URLOpener` does it (e.g., different guessing, etc.), they can copy what the default URLOpener does and modify it how they like, and everything will work from there.

Reply:
I see the merits of this, not expanding struct `bucket` at all and setting `SignBytes` accordingly. Unfortunately the client needs to be closed or its connection remains lingering. I've considered starting a goroutine to close it once the *background context* concludes but found that it might be too late, i. e. the connection lingered still (the goroutine won't block shutdown without anything in place).

Reply:
Further, `SignBytes` does not accept a context, specifically no *request context*, therefore it's not possible to properly express the IAM Client API call properly as merely `SignBytes([]byte)…`.

In extension, the user won't be able to use anything with a *request context* as `SignBytes` without any prior callback or the like to get that context in.

Reply:
If you want I can add `context.Context` as parameter to `SignBytes` and implement your vision for this—that'll be a breaking change, though.

Reply:
I think it's important that the functionality you're providing a default for in the default URLOpener is functionality that's customizable without making changes to the module. I see your point about a `SignBytes` function not being able to access the `context.Context` passed in to `SignedURL`. To resolve that, how about adding a `MakeSignBytes` field to `Options` with the signature `MakeSignBytes(ctx context.Context) func([]byte) ([]byte, error)`? If provided, `SignedURL` would call that function with the request `ctx` to get a `SignBytes` func to pass to `storage.SignedURL` via the `storage.SignedURLOptions`.

Then, your code in the default `URLOpener` would create a `credentialsSigner` client struct, with a lazily initialized client field and a `MakeSignBytes` method, and pass that `MakeSignBytes` method in `Options`.

It's true that the lazily initialized client will never be cleaned up, but that's OK; there's only one of them (not one per bucket), and it will live as long as the application. We've used that pattern in other default URLOpeners (check out `pubsub/rabbitpubsub` for example).

Reply:
I've added a `MakeSignBytes` and expanded the comments accordingly.

Consequently we no longer properly close the client and have lost the ability to impersonate other *service accounts*, but at least the latter can be retrofitted by the user if need be. We've gained the ability to see HSM or TPM modules, and “keyless schemes” (a different process holds the key), be utilized for storing a signing key.